### DISCUSSION TITLE
NRT analysis disable lambdas

### DISCUSSION THREAD

**Discussion by TahirAhmadov** [dotnet/csharplang#6212] [Language Ideas]
With NRT, I think one of the main pitfalls with adoption is the overuse of (evil but necessary) `!` suppress operator. Since it's an easy way to "get rid of the warnings", it diminishes the usefulness of NRT by "cheapening" its importance. Therefore, I think we should strive to make it so that it's possible to write code where `!` is the absolute exception, preferably in a few, well-explained edge cases, as part of official guidelines.
Yet another situation where this comes up is LINQ queries:
```cs
var person = DataAccess.Query<Person>().Where(p=>p.Emails!.Any()).FirstOrDefault();
var company = DataAccess.Query<Company>().Where(c=>c.Address!.City=="Miami").FirstOrDefault();
```
The `Emails` property is a collection, which is not always retrieved for obvious performance reasons, and is marked as nullable to "remind" the consumers that they need to explicitly retrieve it (I raised this in https://github.com/dotnet/csharplang/discussions/6148). The `Address` property is a regular many-to-one relationship, and it's nullable because not all companies have a mailing address on file. In both of these cases, despite the value not being available always, it's perfectly valid to construct a database query using LINQ, because in SQL, it becomes a LEFT JOIN and the WHERE clause will just not return the row - which is expected.
As can be seen above, I keep having to sprinkle `!` in these cases. In a DB-driven application, this is a lot of `!`. To get rid of it, can we introduce an attribute which would be added to `Queryable.Where`, etc., which disables nullability analysis in the `Expression.Lambda<>` that are passed to it? 
Thinking about it, it's possible it's too permissive:
```cs
SomeDto? someDto = null;
var company = DataAccess.Query<Company>().Where(c=>c.Address.City==someDto.City).FirstOrDefault();
```
Here, we got rid of the `!` in the database side, but that means we don't catch the `someDto` being nullable. Maybe it can work so that only the property path starting with lambda parameters have the nullability analysis disabled?

**Comment by PathogenDavid** [comment]
These seem like an improper uses of the damnit operator in the first place. It seems to me like you're forgiving nulls you really shouldn't be.

> The `Emails` property is a collection, which is not always retrieved for obvious performance reasons, and is marked as nullable to "remind" the consumers that they need to explicitly retrieve it

Which means it can be null in otherwise valid situations. I don't see why the compiler should pretend it can't be `null` just because you're accessing it in a lambda expression.

It seems to me like you should just lazily retrieve `Emails` when it isn't present rather than making it the consumer's problem. (Perhaps with a debug assert or something to help discover situations where `Emails` should've been retrieved during the initial query.)

If this sort of thing comes up a lot in your code base, you could consider writing a `DiagnosticSuppressor` to automatically suppress the null warnings automatically in cases where you know it won't be null.

> it's nullable because not all companies have a mailing address on file.

Then why does your predicate assume it isn't null? It seems like you should be doing this instead:

```csharp
var company = DataAccess.Query<Company>().Where(c=>c.Address?.City=="Miami").FirstOrDefault();
```

**Reply by TahirAhmadov** [reply]
> These seem like an improper uses of the damnit operator in the first place. It seems to me like you're forgiving nulls you really shouldn't be.
> 
> > The `Emails` property is a collection, which is not always retrieved for obvious performance reasons, and is marked as nullable to "remind" the consumers that they need to explicitly retrieve it
> 
> Which means it can be null in otherwise valid situations. I don't see why the compiler should pretend it can't be `null` just because you're accessing it in a lambda expression.

It can't be null because that lambda expression is part of the LINQ query, therefore nullability goes out the window. I explained this in the OP, with SQL queries, etc. Excuse me, did you read the OP before writing your reply?

> It seems to me like you should just lazily retrieve `Emails` when it isn't present rather than making it the consumer's problem. (Perhaps with a debug assert or something to help discover situations where `Emails` should've been retrieved during the initial query.)

That is a terrible advice. For anybody else who reads this - do **not** rely on lazy retrieval of collections like that. (The only exception would be in situations where the developer truly has no way of knowing what will be needed - say, an document template placeholder, which needs the email in some templates but not in others.)

> If this sort of thing comes up a lot in your code base, you could consider writing a `DiagnosticSuppressor` to automatically suppress the null warnings automatically in cases where you know it won't be null.

That is a possibility, actually. I'll look into it.

> > it's nullable because not all companies have a mailing address on file.
> 
> Then why does your predicate assume it isn't null? It seems like you should be doing this instead:
> 
> ```cs
> var company = DataAccess.Query<Company>().Where(c=>c.Address?.City=="Miami").FirstOrDefault();
> ```

1, that code doesn't compile - you cannot have null-propagating operator in LINQ expressions. 2, it doesn't really matter, because - like I wrote in the OP - this becomes a SQL query anyway, and LEFT JOIN doesn't care whether you used `!.`, `?.`, or `.`, with or w/o a warning.

**Reply by PathogenDavid** [reply]
> Excuse me, did you read the OP before writing your reply?
> [...]
> That is a terrible advice.

Quite frankly there is no need for such rudeness. There are more productive ways to express you think I've misunderstood you or you don't like what I'm proposing.

> It can't be null because that lambda expression is part of the LINQ query

We've had a bit of a disconnect here. I thought you were using `DataAccess.Query<Person>()` as a stand-in for something more specific which was returning `IEnumerable<Person>`. (In retrospect that doesn't make much sense, but that's how I read it because that's the end of LINQ I normally interact with.)

You aren't dealing with plain ol' lambdas, you're dealing with expression trees. It's no secret that expression trees have rotted over the years. It seems to me like the proper solution here is "NRT analysis should have better support for expression trees / IQueryable", not throwing it out the window entirely.

> For anybody else who reads this - do **not** rely on lazy retrieval of collections like that.

I'm not saying you should *rely* on it. You should try to prevent misuse and have it be a last resort in event of misuse. Another alternative is throwing in the getter if the `Person` was retrieved without emails.

Marking it as nullable when it's only sort-of nullable and then using the damnit operator is just sweeping issues under the rug and turning mistakes into NREs. You're not any better off than when you were using C# 7.

**Reply by TahirAhmadov** [reply]
> I thought you were using DataAccess.Query<Person>() as a stand-in for something more specific which was returning IEnumerable<Person>.

I didn't mean to be rude, but I admit I was quite annoyed because I specifically wrote about LINQ and SQL and LEFT JOIN.

> Another alternative is throwing in the getter if the Person was retrieved without emails.

That's something I considered; I may try doing that, actually, for collections. The downside is, it doesn't give you a compile-time check; you have to run it to catch the problem, and at that point, you might as well just encounter the NRE. Basically, I need to write IL emit logic to override virtual collection properties, and the only benefit would be a more appropriate exception type. Is it worth it? Maybe. **Also,** there is performance overhead of initializing the proxy; if I don't need to use proxies, such as, no lazy loading and no "special collection not loaded exception", I just create the actual POCOs, which can save some real time when getting large lists from the DB (I mean, the hydration itself, not the DB query).

> Marking it as nullable when it's only sort-of nullable and then using the damnit operator is just sweeping issues under the rug and turning mistakes into NREs. You're not any better off than when you were using C# 7.

I agree, partially. Yes, `!` is bad; but, at least I know I wrote the `.Include()` and `!` is justified. It's something, but not perfect; which is why I created #6148; 
However, this discussion of how to retrieve collections belongs in that discussion, not here. This one is more for `Where` lambda expression trees.

> It seems to me like the proper solution here is "NRT analysis should have better support for expression trees / IQueryable", not throwing it out the window entirely.

That's the one part where expression trees currently already work as expected. I know there are a myriad of accumulated shortcomings over the years, including the `?.`, but nullability analysis works just as well as it does in other code. In fact, this proposal aims to improve it by making it slightly different than regular C# code - specifically because the end result is some sort of persistence layer operation, like a SQL query.

**Reply by TahirAhmadov** [reply]
@PathogenDavid Having thought about it, I realized something - perhaps that's what you meant: instead of "turning off" the NRT warnings when navigating on the database side of the lambda, if/when `?.` is implemented in LINQ expressions, it solves the problem without using `!` and in a concise way, too. In fact, I think that may be a much easier solution than the "selective NRT analysis disable" idea I came up with in this discussion.

**Comment by pinkfloydx33** [comment]
I've run into this recently with EF queries and I agree it's a bit annoying. I don't think there's a good way to handle it though. Linq2Objects or just doing AsQueryable() on an enumerable will cause the expression tree to be compiled and evaluated in memory. In that case an exception would be thrown if the property is null. Queryable/ExpressionTrees aren't limited to SQL and that's going to be a problem.

The solution is likely a diagnostic suppressor that looks for annotations/attributes on the query root expression (Dbset in EF or your Query call) and then suppresses analysis inside the lambda, possibly based on sub-expressions of the ParameterExpression, but even that might need some bottom-up analysis. Unfortunately that would still fall apart the second you pass the queryable around and are treating it as a generic IQueryable (and not something rooted from a specific source) 

I'm pretty sure there's similar discussions in the EF repo and that they were exploring solutions in this area. Perhaps whatever it is could be made generic/multipurpose.

**Reply by TahirAhmadov** [reply]
I agree, it's complicated. I'm not insisting on any specific solution; for me, the equally important thing is to raise the question and solicit proposals from others. Your idea of the root expression attribute is a solid idea.
